prereise/gather/griddata/hifld/data_process/generators.py

```diff
@@ -22,54 +24,108 @@ def floatify(value, default=float("nan")):
-            return matching_subs.index[0]
-        except KeyError:
-            return pd.NA
-    try:
-        # This ZIP code contains substations, this block will execute successfully
-        matching_subs = substation_groupby.get_group(lookup_params)
-    except KeyError:
-        # If this ZIP code does not contain substations, this block will execute, and
-        # we select a set of 'nearby' substations
-        zip_range = [int(generator.loc["ZIP"]) + offset for offset in range(-100, 101)]
-        zip_range_strings = [str(z).rjust(5, "0") for z in zip_range]
-        try:
-            matching_subs = pd.concat(
-                [
-                    substation_groupby.get_group((generator.loc["interconnect"], z))
-                    for z in zip_range_strings
-                    if (generator.loc["interconnect"], z) in substation_groupby.groups
-                ]
-            )
-        except ValueError:
-            # If no matching subs within the given interconnection and ZIPs, give up
+
+    def get_sub_id_of_closest_substation(generator, state_trees, subs_state_lookup):
+        if not isinstance(generator["xyz"], list):
             return pd.NA
-    distance_to_subs = matching_subs.apply(
-        lambda x: haversine((x.lat, x.lon), (generator.lat, generator.lon)),
+        grouper_key = (generator["interconnect"], generator["state"])
+        _, array_index = state_trees[grouper_key].query(generator["xyz"])
+        sub_index = subs_state_lookup[grouper_key][array_index]
+        return sub_index
+
+    # Translate lat/lon to 3D positions (assume spherical earth, origin at center)
+    substations_with_xyz = substations.assign(
+        xyz=substations.apply(lambda x: latlon_to_xyz(x["lat"], x["lon"]), axis=1)
+    )
+    generators_with_xyz = generators.assign(
+        xyz=generators.apply(
+            lambda x: (
+                pd.NA
+                if pd.isna(x["lat"]) or pd.isna(x["lon"])
+                else latlon_to_xyz(x["lat"], x["lon"])
+            ),
+            axis=1,
+        )
+    )
+
+    # Group substations by state to build KDTrees
+    subs_state_lookup = substations_with_xyz.groupby(["interconnect", "STATE"]).groups
+    # Group substations by ZIP code for a fallback for generators without coordinates
+    subs_zip_groupby = substations_with_xyz.groupby(["interconnect", "ZIP"])
+
+    # Create a KDTree for each combination of state and interconnect
+    state_trees = {
+        key: KDTree(np.array(substations_with_xyz.loc[sub_ids, "xyz"].tolist()))
+        for key, sub_ids in subs_state_lookup.items()
+    }
+    # Ensure that we have a tree for every generator
+    gens_state_groupby = generators_with_xyz.groupby(["interconnect", "state"])
+    missing_groups = set(gens_state_groupby.groups) - set(state_trees)
+    if len(missing_groups) > 0:
+        # There are some combinations of generator (interconnect, state) without subs
+        allowable_border_states = {"KS", "NE", "OK"}
```

Comment:
Are these states coming from EIA references?

Reply:
State information for generators comes from the EIA. The 'allowable' border states come from my personal eyeballing. I should probably add South Dakota too.